conda_package_streaming/conda_fmt.py

```diff
@@ -0,0 +1,179 @@
+"""
+Tools for creating ``.conda``-format archives.
+
+Uses ``tempfile.SpooledTemporaryFile`` to buffer ``pkg-*.tar`` and
+``info-*.tar``, then compress directly into an open `ZipFile` at the end.
+`SpooledTemporaryFile` buffers the first 10MB of the package and its metadata in
+memory, but writes out to disk for larger packages.
+
+Uses more disk space than ``conda-package-handling`` (temporary uncompressed
+tarballs of the package contents) but accepts streams instead of just
+files-on-the-filesystem.
+"""
+
+from __future__ import annotations
+
+import json
+import shutil
+import tarfile
+import tempfile
+import zipfile
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Callable, Iterator
+
+import zstandard
+
+# increase to reduce speed and increase compression (levels above 19 use much
+# more memory)
+ZSTD_COMPRESS_LEVEL = 19
+# increase to reduce compression and increase speed
+ZSTD_COMPRESS_THREADS = 1
+
+CONDA_PACKAGE_FORMAT_VERSION = 2
+
+# Account for growth from "2 GB of /dev/urandom" to not exceed ZIP64_LIMIT after
+# compression
+CONDA_ZIP64_LIMIT = zipfile.ZIP64_LIMIT - (1 << 18) - 1
+
+
+def anonymize(tarinfo: tarfile.TarInfo):
+    """
+    Pass to ``tarfile.add(..., filter=anonymize)`` to anonymize uid/gid.
+
+    Does not anonymize mtime or any other field.
+    """
+    tarinfo.uid = tarinfo.gid = 0
+    tarinfo.uname = tarinfo.gname = ""
+    return tarinfo
+
+
+class CondaTarFile(tarfile.TarFile):
+    """
+    Subclass of :external+python:py:class:`tarfile.TarFile` that adds members to
+    a second ``info`` tar if they match ``is_info(name)``.
+
+    Create this with ``conda_builder(...)`` which sets up the component
+    archives, then wraps them into a ``.conda`` on exit.
+
+    Only useful for creating, not extracting ``.conda``.
+    """
+
+    info_tar: tarfile.TarFile
+    is_info: Callable
+
+    def __init__(
+        self,
+        *args,
+        info_tar: tarfile.TarFile,
+        is_info=lambda name: name.startswith("info/"),
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.info_tar = info_tar
+        self.is_info = is_info
+
+    def addfile(self, tarinfo, fileobj=None):
```

Comment:
Or we could delegate to two other TarFile instances (info, pkg) and never call self.addfile()

Reply:
I think two TarFile instances are preferrable, even if a slight memory overhead it would reduce the risk of introducing some file access edge case with a custom tarfile subclass.

Reply:
File access?

Reply:
Okay, I don't have strong feelings about it